%YAML:1.0

#Multiple thread support
multiple_thread: 0

# segmmentation
segment_cloud: 1
horizon_scan: 400
min_cluster_size: 15
min_line_size: 10
segment_valid_point_num: 5
segment_valid_line_num: 3
segment_theta: 0.53 # Lego-loam: 1.047

# laser parameters
idx_ref: 0
n_scans: 16
num_of_laser: 2

cloud_topic:
   - "/left/velodyne_points"
   - "/right/velodyne_points"

# window sizes
window_size: 6
opt_window_size: 3

# Extrinsic parameter between IMU and Camera.
estimate_extrinsic: 1   # 0  Have an accurate extrinsic parameters. We will trust the following imu^R_cam, imu^T_cam, don't change it.
                        # 1  Have an initial guess about extrinsic parameters. We will optimize around your initial guess.
                        # 2  Have no prior about extrinsic parameters. We will initialize and optimize around them

body_T_laser: !!opencv-matrix
   rows: 2
   cols: 7
   dt: d
   # data: [0, 0, 0, 1, 0, 0, 0,
   #        0, 0, 0, 1, 0, 0, 0] # qx qy qz qw px py pz
   data: [0.00000, 0.00000, 0.00000, 1.00000, 0.00000, 0.00000, 0.00000,
          0.31, 0.0, 0.0, 0.94, 0, -0.2, 0.0] # initialization   
   # data: [0.00000, 0.00000, 0.00000, 1.00000, 0.00000, 0.00000, 0.00000,
   #        0.32597, -0.0009444, -0.000578, 0.94538, 0.003173, -0.49107, -0.214356] # estimate         
   # data: [0.00000, 0.00000, 0.00000, 1.00000, 0.00000, 0.00000, 0.00000,
   #        0.342, 0.0, 0.0, 0.940, 0, -0.477, -0.220] # gt

   # <!-- laser_0 to laser_1 -->
   # <!-- 0.342, 0.0, 0.0, 0.940, 0, -0.477, -0.220 -->
   # <!-- base_link to laser_0 -->
   # <!-- -0.216 0 0 0.976 0 0.266 0.734 --> 25 degrees
   # <!-- base_link to laser_1 -->
   # <!-- 0.131 0 0 0.991 0 -0.260 0.736 --> 15 degrees

#unsynchronization parameters
estimate_td: 0                      # 0 have an accurate time offset among sensors
                                    # 1 online estimate time offset among sensors

td: !!opencv-matrix                 # initial value of time offset. unit: s. readed image clock + td = real image clock (IMU clock)
   rows: 1
   cols: 2
   dt: d
   data: [0, 0]

distortion: 0
scan_period: 0.1
laser_sync_threshold: 0.07

######################################################## odometry
# optimization 
max_solver_time: 0.02  # max solver itration time (s), to guarantee real time
max_num_iterations: 5   # max solver itrations, to guarantee real time

roi_range: 0.5
distance_sq_threshold: 25
nearby_scan: 2.5

# movement type
planar_movement: 1

# feature corresponding paprameters
min_match_sq_dis: 1.0
min_plane_dis: 0.2

# factor
marginalization_factor: 1
point_plane_factor: 1
point_edge_factor: 1
prior_factor: 1
prior_factor_pos: 0
prior_factor_rot: 5 

pcl_viewer: 0
pcl_viewer_normal_ratio: 10

# calibration converage parameters
n_cumu_feature: 10
eig_initial: 100
eig_thre_calib: 10000000
n_calib: 20

# good feature
odom_gf_ratio: 1.0

skip_num_odom_pub: 1

######################################################## mapping
map_corner_res: 0.2
map_surf_res: 0.4
map_outlier_res: 0.8
map_eig_thre: 50

# uncertainty parameter
# translation, rotation, point (rho, phi, zeta)
uct_ext: !!opencv-matrix
   rows: 2
   cols: 6
   dt: d
   data: [0, 0, 0, 0, 0, 0,
          0.0025, 0.0025, 0.0025, 0.01, 0.01, 0.01] 

uct_measurement: !!opencv-matrix
   rows: 1
   cols: 3
   dt: d
   data: [0.0025, 0.0025, 0.0025]

trace_threshold_before_mapping: 2.0
trace_threshold_after_mapping: 2.0
